ddtrace/internal/rate_limiter.py

```diff
@@ -38,54 +38,57 @@ def __init__(self, rate_limit):
         self.tokens = rate_limit  # type: float
         self.max_tokens = rate_limit
 
-        self.last_update = compat.monotonic()
+        self.last_update_ns = compat.monotonic_ns()
 
-        self.current_window = 0  # type: float
+        self.current_window_ns = 0  # type: float
         self.tokens_allowed = 0
         self.tokens_total = 0
         self.prev_window_rate = None  # type: Optional[float]
 
         self._lock = threading.Lock()
 
-    def is_allowed(self):
-        # type: () -> bool
+    def is_allowed(self, timestamp_ns=None):
+        # type: (Optional[int]) -> bool
         """
         Check whether the current request is allowed or not
 
         This method will also reduce the number of available tokens by 1
 
+        :param int timestamp_ns: Optional timestamp in nanoseconds for the current request.
+            If not provided `compat.monotonic_ns()` is used.
         :returns: Whether the current request is allowed or not
         :rtype: :obj:`bool`
         """
+        if timestamp_ns is None:
+            timestamp_ns = compat.monotonic_ns()
         # Determine if it is allowed
-        allowed = self._is_allowed()
+        allowed = self._is_allowed(timestamp_ns)
         # Update counts used to determine effective rate
-        self._update_rate_counts(allowed)
+        self._update_rate_counts(allowed, timestamp_ns)
         return allowed
 
-    def _update_rate_counts(self, allowed):
-        # type: (bool) -> None
-        now = compat.monotonic()
-
+    def _update_rate_counts(self, allowed, timestamp_ns):
+        # type: (bool, int) -> None
         # No tokens have been seen yet, start a new window
-        if not self.current_window:
-            self.current_window = now
+        if not self.current_window_ns:
+            self.current_window_ns = timestamp_ns
 
         # If more than 1 second has past since last window, reset
-        elif now - self.current_window >= 1.0:
+        # DEV: We are comparing nanoseconds, so 1e9 is 1 second
+        elif timestamp_ns - self.current_window_ns >= 1e9:
             # Store previous window's rate to average with current for `.effective_rate`
             self.prev_window_rate = self._current_window_rate()
             self.tokens_allowed = 0
             self.tokens_total = 0
-            self.current_window = now
+            self.current_window_ns = timestamp_ns
```

Comment:
We don't handle the `else` here but it could happen with a custom span start (or also because the span time isn't monotonic). Probably at least worth handling and logging if it happens. Should probably be a warning log since it would be more likely to occur with a custom span start than a clock discrepancy.

Reply:
the `if`/`elif` cases are just if we need to change the current window, the else case is not changing the current window.

then the final piece is updating the token counts for the current window

Reply:
:+1: resolving since custom span starts aren't possible (as discussed below)